### PR TITLE
Add test for `traverseDoc` and `findInDoc`

### PR DESCRIPTION
**Pull request by fisker**
## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/document/utils.js [hunk 1]
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
   }
 }
 
-function findInDoc(doc, fn, defaultValue) {
-  let result = defaultValue;
-  let hasStopped = false;
```
Comment (fisker) [inline src/document/utils.js:163]: This is useless, `onEnter` returns `false` will stop traverse, must left over during refactoring https://github.com/prettier/prettier/pull/442

File: src/document/utils.js [hunk 2]
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
...
-  function findInDocOnEnterFn(doc) {
+function findInDoc(doc, fn) {
+  let result;
+  traverseDoc(doc, (doc) => {
     const maybeResult = fn(doc);
-    if (maybeResult !== undefined) {
```
Comment (fisker) [inline src/document/utils.js:166]: I don't think we want use `findInDoc` to do find falsely value.

File: src/document/utils.js [hunk 3]
```diff
@@ -158,37 +158,34 @@ function mapDoc(doc, cb) {
...
+        (doc.type === DOC_TYPE_LINE && doc.hard) ||
+        doc.type === DOC_TYPE_BREAK_PARENT
+    )
+  );
 }
 
-function willBreak(doc) {
-  return findInDoc(doc, willBreakFn, false);
+function canBreak(doc) {
+  return Boolean(findInDoc(doc, (doc) => doc.type === DOC_TYPE_LINE));
```
Comment (fisker) [inline src/document/utils.js:188]: This not seems correct to me, line may live in a unbreakable group. But it's out scope of this PR.

File: src/document/utils.js (hunk 1 above)
Comment (fisker) [reply src/document/utils.js:163]: Maybe I'm wrong, adding test.

Comment (fisker) [reply src/document/utils.js:163]: return false on `onEnter` doesn't stop traverse. https://github.com/prettier/prettier/pull/13349/commits/e3ff2edc6f778797659dbd8a323a4d26a7be3405

**Comment by thorn0** [comment]
The named functions were used to make profiling more convenient as dev tools show those names in reports. I know we don't profile Prettier much these days, but if we need to do that, we might end up having to return these names

**Comment by thorn0** [comment]
Overall, it's already a very simple function. Also it's exported and may be used by plugins. I'd just remove `hasStopped` and call it a day.

File: src/document/utils.js [hunk 4]
```diff
@@ -161,16 +161,20 @@ function mapDoc(doc, cb) {
 function findInDoc(doc, fn, defaultValue) {
   let result = defaultValue;
   let hasStopped = false;
+
   function findInDocOnEnterFn(doc) {
+    if (hasStopped) {
+      // Stop processing rest docs
+      return false;
```
Comment (thorn0) [inline src/document/utils.js:168]: Doesn't `return false` stop the further processing immediately? Is the `hasStopped` variable really needed?

**Comment by fisker** [comment]
Okay, I reverted change, so this PR only going to add some tests.

A minor improvement, the `findInDoc` previously keep running function on the remaining docs, but it should not. I moved `hasStopped` check to top.

File: src/document/utils.js (hunk 4 above)
Comment (fisker) [reply src/document/utils.js:168]: See https://github.com/prettier/prettier/pull/13349#discussion_r952138588

Comment (fisker) [reply src/document/utils.js:168]: I thought it does, but it's not. It will keep processing the siblings.

Comment (thorn0) [reply src/document/utils.js:168]: Ah, okay. I see it's needed.

File: src/document/utils.js [hunk 5]
```diff
@@ -161,16 +161,20 @@ function mapDoc(doc, cb) {
 function findInDoc(doc, fn, defaultValue) {
   let result = defaultValue;
   let hasStopped = false;
+
   function findInDocOnEnterFn(doc) {
+    if (hasStopped) {
+      // Stop processing rest docs
```
Comment (thorn0) [inline src/document/utils.js:167]: Yes, the top seems to be a more correct place. Let's just not use the word "stop" in the comment:
```suggestion
      // Skip further processing
```

**Comment by thorn0** [comment]
How about renaming `hasStopped` to `shouldSkipFurtherProcessing` and removing the comment?

**Pull request merged**